engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -4720,8 +4721,11 @@ protected VirtualMachine retrieve() {
         final User user = context.getCallingUser();
         final Account account = context.getCallingAccount();
 
-        for (StoragePool destPool : dest.getStorageForDisks().values()) {
-            checkConcurrentJobsPerDatastoreThreshhold(destPool);
+        Collection<StoragePool> storagePools = dest.getStorageForDisks().values();
```

Comment:
so how about when dest or dest.getStorageForDisks() is null, this would still give a npe, no?

Reply:
That's true, the nullity check has to be for the Map returned by 'dest.getStorageForDisks()'.
Refactored to use this approach;

        Map<Volume, StoragePool> volumeStorageMap = dest.getStorageForDisks();
        if (volumeStorageMap != null) {
            for (Volume vol : volumeStorageMap.keySet()) {
                checkConcurrentJobsPerDatastoreThreshhold(volumeStorageMap.get(vol));
            }
        }